File: 234-palindrome-linked-list/palindrome-linked-list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* reverseLL(ListNode* head){
        if(head == nullptr || head-> next == nullptr)
            return head;
        ListNode* temp = head;
        ListNode* prev = nullptr;
        while(temp != nullptr){
            ListNode* front = temp -> next;
            temp -> next = prev;
            prev = temp;
            temp = front;
        }
        return prev;
    }

    bool isPalindrome(ListNode* head) {
        ListNode* slow = head;
        ListNode* fast = head;
        while(fast -> next != nullptr && fast -> next -> next != nullptr){
            slow = slow -> next;
            fast = fast -> next -> next;
        }
        ListNode* newHead = reverseLL(slow -> next);
        ListNode* first = head;
        ListNode* second = newHead;
        while(second != nullptr){
            if(first -> val != second -> val){
                reverseLL(newHead);
                return false;
            }
            first = first -> next;
            second = second -> next;
        }
        reverseLL(newHead);
        return true;
    }
};

File: 234-palindrome-linked-list/palindrome-linked-list.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} head
 * @return {boolean}
 */

var reverseLL = function (head) {
    if (!head || !head.next)
        return head;

    let temp = head;
    let prev = null;
    while (temp) {
        front = temp.next;
        temp.next = prev;
        prev = temp;
        temp = front;
    }
    return prev;
};

var isPalindrome = function (head) {
    if (!head || !head.next)
        return true;

    let slow = head, fast = head;
    while (fast.next !== null && fast.next.next !== null) {
        slow = slow.next;
        fast = fast.next.next;
    }

    let newHead = reverseLL(slow.next);
    let first = head, second = newHead;

    while (second) {
        if (first.val !== second.val) {
            reverseLL(newHead);
            return false;
        }

        first = first.next;
        second = second.next;
    }

    reverseLL(newHead);
    return true;
};